Add-on modules in Studio Pro (#4005)
mendix/docs: content/refguide/add-on-module.md

---
title: "Add-On Module"
parent: "modules"
menu_order: 10
tags: ["studio pro", "protected module", "module"]
---

## 1 Introduction

> An **add-on** module is a special kind of Marketplace module that adds functionality to your app. 

Reviewer: It's not necessarily a Marketplace module. The `.mxmodule` file can be distributed by hand, just like a traditional `.mpk` file of a template, module or widget.

Author: Removed the word "Marketplace".

> An **add-on** module is a special kind of the Marketplace module that adds functionality to your app. 

Reviewer 2: No "the"

Author: I don't remember the grammar rule from the top of my mind, but "the" should be there.

Reviewer 2: I also don't know the grammar rule but I know intuitively that there can be no "the" there. Please check with a native speaker.

Author: I know it should be there. You can see that we use "the" for module names, property names, etc. in our documentation.

Author: I missed that there is "a special kind" in front. That changes it. Thanks, Arjan!

> Add-on modules may have Intellectual Property (IP) protection enabled and have the `.mxmodule` extension. For more information on IP protection, see [IP Protection](/appstore/creating-content/sol-ip-protection). 

Reviewer: > [...] may have [IP] protection enabled

This might be slightly misleading. In the first version of Add-on modules, all model content is hidden, except for the usable documents, and e.g. for microflows and java actions, even their contents are hidden. I'm OK with keeping it as-is, but please be aware of this in the docs team :-)

Author: I removed the word "may" to indicate that it is not something optional.

> Add-on modules may have Intellectual Property (IP) protection enabled and have an extension `.mxmodule`. For more information on IP protection, see [IP Protection](/appstore/creating-content/sol-ip-protection). 

Reviewer 2: an -> the

> An example of the add-on module is the **MS Teams Connector** module. {{% todo %}}[Add a link to the MS Teams Connector document.]{{% /todo %}} 

Reviewer 2: the -> an

> When you download it, it shows up in a new **Add-ons** folder in the App Explorer. The main difference of the add-on module is that some functionality and elements, such as microflows and entities, are protected and, therefore, *hidden* from other users by the publisher. This means that you cannot see how the microflow is built, but you are able to use it in your app, for example, you can call the microflow with protected implementation from another microflow:

Reviewer 2: Perhaps you can stress that this also means that an add-on module has a clear API: you know which things to use, namely the visible things.

Author: Added

> When you download it, it shows up in a new **Add-ons** folder in the App Explorer. The main difference of the add-on module is that some functionality and elements, such as microflows and entities, are protected and, therefore, *hidden* from other users by the publisher. This means that add-on modules has a clear API: you know which elements you can use. For example, you cannot see how the microflow is built, but you are able to use it in your app, for example, you can call the microflow with protected implementation from another microflow:

Reviewer: > [...] are protected and [...]

I suggest to change this to:

> the implementation details are hidden from consumers so that IP is protected.

Author: Done

> When you download it, it shows up in a new **Add-ons** folder in the App Explorer. The main difference of the add-on module is that some functionality and elements, such as microflows and entities, are protected and, therefore, *hidden* from other users by the publisher. This means that add-on modules has a clear API: you know which elements you can use. For example, you cannot see how the microflow is built, but you are able to use it in your app, for example, you can call the microflow with protected implementation from another microflow:

Reviewer: > This means that add-on modules has a clear API: you know which elements you can use.

Add:

> When you open the microflow, you will see its API including the microflow and parameter documentation.

Author: Rewritten

![Protected Microflow Example](attachments/add-on-module/protected-microflow.png)

> You also might not see the full domain model of the module, but entities, attributes, and associations that were specifically exposed by the published. 

Reviewer 2: but -> but only

Reviewer 2: published -> publisher

Author: Good catch with "publisher"! 
"but" here is used not as a conjunction, but as a preposition and a synonym to "except for", in this case we cannot substitute it with "but only". However, I might change it to simply "only" or "except for", they might be less confusing.

> You also might not see the full domain model of the module, only entities, attributes, and associations that were specifically exposed by the publisher. 

Reviewer: > [...] by the publisher

Add: The visible parts of the domain model are intended to be reused. Whatever might be hidden is not needed for proper use of the module.

Author: Done

{{% alert type="warning" %}}

> Mendix Studio does not support add-on modules, you will not be able to open the app that uses them in Studio.

Reviewer: ```suggestion
Mendix Studio does not currently support add-on modules, you will not be able to open any apps in Studio that use add-on modules.
```

{{% /alert %}}

## 2 Importing an Add-On Module to Your App

For information on how to import a module in your app, see the [Adding a Module](/appstore/app-store-content#add-module) section in *Use Marketplace Content in Studio Pro*.

> {{% todo %}}[Double-check which import option to leave.]{{% /todo %}} 

Author: I will leave the description below for now. I see that MS Teams Connector team is describing this option in [Luyao's PR](https://github.com/mendix/docs/pull/3678).

To import an add-on module, do the following:

> 1. Download the add-on module from the Marketplace.

Reviewer 2: As discussed in Slack, you can simply download the module from the marketplace that is built into Studio Pro while having your app open. Then you do not need steps 3 up to and including 6.

Author: @ironthorn , but the MS Team Connector team have quite the opposite in the document they are preparing on the connector. That is why I left this option to double-check before the release.

Author: Agreed with David that I add a cross-reference to another document and MS Teams Connector document will include instructions specific for this component.

2. Open the app in Studio Pro where you would like to import this module.
3. Go to **File** > **Import Module Package** or right-click the app name in the App Explorer and select **Import module package**.
4. Browse to the downloaded module package and click **Open**.
5. In the **Import Module** pop-up window, select **Add as new module** in **Action**.
6. Go to **App** > **Synchronize App Directory**.

You have imported the add-on module.  

## 3 Updating an Add-On Module 

For information on how to update a module in your app, see the [Updating a Module](/appstore/general/app-store-content#update-module) section in *Use Marketplace Content in Studio Pro*.

## 4 Deleting an Add-On Module

To delete an add-on module, do the following:

1. Go to  **App** > **Show App Directory in Explorer** to open the app folder.

> 2. Browse to the **modules** folder and remove the ZIP file from there.

Reviewer 2: ZIP file -> mxmodule file
